NVIDIA/NeMo-Guardrails
# build2023/switch to pyproject

Hello there! Finally had time to work on #234. I realised that to fully address the issue, I need to first switch the previous settings to a pyproject.toml-based configuration, then start working on the constraints.

## Changes

1. Removed the following:

- requirements.txt
- requirements-dev.txt
- MANIFEST.in
- pytest.ini
- setup.py

And replaced their contents with a single `pyproject.toml` file, following these guides:

- https://packaging.python.org/en/latest/guides/writing-pyproject-toml/
- https://setuptools.pypa.io/en/stable/userguide/pyproject_config.html

2.  Modified the CONTRIBUTING.md file accordingly. Now the contributor just needs to run the following line:

```bash
[python -m] pip install ".[dev]"
```

3. Configured the `version` property to be read directly from `nemoguardrdails.__version__` so you just have to change it in one place.
4. Removed `setuptools` from the dependency list.
5. Moved `watchdog` to core dependencies.

## How to test

All tests run successfully, yay!

To test whether the new setup works, we should push a release to TestPyPI. I don't know what [schema](https://packaging.python.org/en/latest/specifications/version-specifiers/#version-scheme) you should use. Might just be a dev release like `0.6.1.dev<whatever>` since it's just a throwaway.

## Questions

1. Why do you package the following non-python files?

```
  "**/*.yml",
  "**/*.co",
  "**/*.txt",
  "**/*.json",
  "../examples/**/*",
  "../chat-ui/**/*",
``` 

2. Would you like to separate the CLI dependencies and make them optional, e.g. `pip install nemoguardrails[cli]`? This would allow to remove starlette, watchdog, fastapi, uvicorn and typer out of core dependencies.
3. Why are you using both aiohttp and httpx?

## Next steps

1. I could move pylint configs under pyproject.toml if you want.
2. Would you like switching from black + isort to ruff? It's blazingly fast. It can replace some functionalities of pylint too.
3. I would try to relax every constraint once at a time and see whether tests fail or not. If they do, I would just place a `<=` constraint for that lib. Do you think this might work?
4. Think of a different package manager. **The TLDR is: if you don't fell like adding new dependencies in the near future, or don't mind manually editing them in pyproject.toml, we can leave it like this**. The long story is in the toggle menu below.

<details>
<summary>Packaging info</summary>

Python PEP 517 and 518 specify the behaviours of packaging _frontends_ and _backends_. Backends like setuptools are what builds the wheels behind the curtains. Frontends usually provide a unified interface (see [here](https://chriswarrick.com/blog/2023/01/15/how-to-improve-python-packaging/)) to work with backends, publish packages, manage virtualenvs and other amenities. Frontends that comply with PEPs will be compatibile with _any_ backend.

To the best of my knowledge, PDM is the most feature-complete frontend that complies with PEPs. But this project does not really need to make a switch if the dependencies won't change much in the near future.

This is why I left setuptools as backend. This means the project can be managed as you used to: just add new dependencies _by hand_ to the `pyproject.toml`'s `dependencies` or `optional-dependencies` table(s).

In the future, if using pip + venv + setuptools + twine (to publish the package) is too much, you can always switch to PDM. It will require some small changes to pyproject.toml, but not many.
</details>
